### PR TITLE
DataLinks: Add DataLinks support to Gauge, BarGauge and SingleStat2 panel

### PR DESCRIPTION
**Pull request by ryantxu**
This PR adds DataLink to FieldConfig and calculates the LinkModel within DisplayValues.

The goal is to have a standard way to represent click behavior across all panels.

TODO:
- [x] implement the click behavior in Gauge panel :)
- [x] improve the method signature for passing the clicked object.  Likely it should include the `row` info
- [x] Sync the UX style for 'Add Mapping' vs 'Add Data Link':
image
- [x] DataLinksContext menu component that wraps visualization and handles datalinks

### CODE REVIEW

**Comment by dprokop** [comment]
` Sync the UX style for 'Add Mapping' vs 'Add Data Link'`

IMO we should keep the Data links approach where the actionable element is within the content of the "panel". I don't think we should spend much time on this as we are about to rethink the options  UI fairly soon :)

File: packages/grafana-ui/src/utils/fieldDisplay.test.ts [hunk 1]
```diff
@@ -154,4 +176,62 @@ describe('FieldDisplay', () => {
     const display = getFieldDisplayValues(options);
     expect(display[0].field.thresholds!.length).toEqual(1);
   });
+
+  it('Should pass value time to linker function via scoped vars', () => {
+    const linkerSpy = jest.fn();
+    const linkerMock: FieldDisplayLinkFunction = (options: FieldDisplayLinkOptions) => {
+      linkerSpy(options.scopedVars);
+      // console.log(options.scopedVars)
```
Comment (dprokop) [inline packages/grafana-ui/src/utils/fieldDisplay.test.ts]:
```suggestion
```

**Comment by dprokop** [comment]
@ryantxu added value time passing into data links here via scoped vars. Sorry for two commits with the same message. Also I have removed the datapoint optional dependency in getDataLinkUIModel. Now anything that can be interpolated in data link should be passed via scoped vars (as updated in graph.ts).

File: packages/grafana-ui/src/utils/fieldDisplay.ts [hunk 2]
```diff
@@ -63,20 +67,42 @@ export interface FieldDisplay {
   field: FieldConfig;
   display: DisplayValue;
   sparkline?: GraphSeriesValue[][];
+  links?: LinkModel[]; // Links with the values escaped
```
Comment (torkelo) [inline packages/grafana-ui/src/utils/fieldDisplay.ts]: Stupid idea maybe, but one potentially huge performance improvement would be to make this lazy, so link interpolation & building the final URL is done on when needed.

**Comment by dprokop** [comment]
@ryantxu experimented with some UI for Gauge panel (cursor indicating there's a data link available) that would work in different visualisations. Landed with this solution that shows link icon next to the cursor. Probably not the best one though.

**Comment by dprokop** [comment]
Added data links to single stat, not very happy with code reusability, need to tweak that a bit

File: packages/grafana-ui/src/utils/fieldDisplay.ts (hunk 2 above)
Comment (dprokop) [reply packages/grafana-ui/src/utils/fieldDisplay.ts]: Changed that in 5887de4

Comment (ryantxu) [reply packages/grafana-ui/src/utils/fieldDisplay.ts]: I removed it entirely from FieldDisplay... and then only create it on demand when the context menu is opened

**Comment by ryantxu** [comment]
@dprokop -- I make some kinda drastic changes.  I want `click` actions to have access to full field and row objects so they could broadcast them. So this adds:
```
  // Expose to the original values for delayed inspection (DataLinks etc)
  view?: DataFrameView;
  column?: number; // The field column index
  row?: number; // only filled in when the value is from a row (ie, not a reduction)
```
with these three options, we have access to everything we could want


Rather than having link creation in DisplayField, it passes DisplayField to LinkService and asks for a supplier.  That should:
1. delay link creation as long as possible
2. keep link creation implemented near the nitty-gritty details

**Comment by dprokop** [comment]
@ryantxu I like the idea of the suppliers! But I felt like adding view/column/row to FieldDisplay isn't enough to support actions that are not necessarily redirects. That's why I I have slightly modified LinkModel to contain an origin of the interaction as well:

```
interface LinkModel<T> {
  ...,
  origin: T;
}
```

Thanks to that it's up to the UI component to decide what to do with the provided link, either use href from it OR utilise the data prom the origin.

Also, I have slightly refactored the `LinkSrv.getLinkSupplier` to be an util instead of a linkSrv member https://github.com/grafana/grafana/pull/18605/files#diff-172511ebd9d4d5c6a4dbf7e82f85c83fR55 and created panel and field links suppliers. Not sure we should use `getLinkSupplier` to decide which supplier to use based on value type. Don't like the type guards there to decide which supplier to use, I have also encuntered some TS problems with this approach that I couldn't resolve. Don't have a strong opinion on that though, WDYT @ryantxu @torkelo ?

**Comment by dprokop** [comment]
@ryantxu any suggestions how to best get series name/value time from DataFrameView?

**Comment by ryantxu** [comment]
@dprokop -- view has a property to get the underlying `.dataFrame` -- then you can find it there.  Note that time only makes sense if there is a row & column specified otherwise the value is likely the result of some reducer (like the mean value)

File: packages/grafana-ui/src/components/BarGauge/BarGauge.tsx [hunk 3]
```diff
@@ -26,7 +26,7 @@ export interface Props extends Themeable {
   orientation: VizOrientation;
   itemSpacing?: number;
   displayMode: 'basic' | 'lcd' | 'gradient';
-  links?: LinkModelSupplier; // only exists if Links Exist
+  links?: LinkModelSupplier<FieldDisplay>; // only exists if Links Exist
```
[the remark this answers is not in the document]
Comment (dprokop) [reply packages/grafana-ui/src/components/BarGauge/BarGauge.tsx]: If you wanna handle links not to be href based (onClick handler for instance), then it's gonna be helpful to know the underlying origin type.

[the remark this answers is not in the document]
Comment (dprokop) [reply packages/grafana-ui/src/components/BarGauge/BarGauge.tsx]: As discussed, about to remove links from the visualization. In current case, with DataLinks being available via context menu, we can handle that outside of the vis.

**Comment by dprokop** [comment]
@torkelo @ryantxu made a shared component for displaying data links as context menu in Gauge/BarGauge and SingleStat2 component. Decided to get rid of the custom cursor indicating there's a context menu available in favour `cursor: context-menu` (vide https://www.w3schools.com/cssref/pr_class_cursor.asp). I think we should strive to review and test this soon and start working on improvements in a separate PR, as this one is growing too big

**Review by ryantxu — COMMENTED** [review]
@dprokop -- this is looking really great.  It is fun to see things oscillate around a solution and settle on something pretty elegant.

Yes lots more to do, and I agree the next changes should be in their own issues.


note: I am not able to approve the request since I made it... but I would if I could :)

**Comment by torkelo** [comment]
Agee, these part changes are a big improvement!

File: public/app/features/panel/panellinks/linkSuppliers.ts [hunk 4]
```diff
@@ -39,7 +52,7 @@ const getFieldLinksSupplier = (value: FieldDisplay): LinkModelSupplier<FieldDisp
 export const getPanelLinksSupplier = (value: PanelModel): LinkModelSupplier<PanelModel> => {
   const links = value.links;
 
-  if (!links || !links.length) {
+  if (!links || (links && links.length === 0)) {
```
[the remark this answers is not in the document]
Comment (torkelo) [reply public/app/features/panel/panellinks/linkSuppliers.ts]: if links is undefined or null the first condition is true, so the other condition should not need to check if links is defined

[the remark this answers is not in the document]
Comment (dprokop) [reply public/app/features/panel/panellinks/linkSuppliers.ts]: Yes, not sure why I've change that ;)

File: packages/grafana-ui/src/components/BarGauge/BarGauge.tsx [hunk 5]
```diff
@@ -26,7 +26,8 @@ export interface Props extends Themeable {
   orientation: VizOrientation;
   itemSpacing?: number;
   displayMode: 'basic' | 'lcd' | 'gradient';
-  links?: LinkModelSupplier<FieldDisplay>; // only exists if Links Exist
+  onClick?: React.MouseEventHandler<HTMLElement>;
+  className?: string;
```
[the remark this answers is not in the document]
Comment (dprokop) [reply packages/grafana-ui/src/components/BarGauge/BarGauge.tsx]: Good point. `className` makes sense in Gauge context, as it's added to the top level dom element in that component. But in this case it's added down the hierarchy. I like the `clickTargetClassName`.

**Review by dprokop — APPROVED** [review]

**Pull request merged**